src/scripts/deactivate-user.ts

```diff
@@ -1,43 +1,40 @@
-import { UserStatus, User as OktaUser } from '@okta/okta-sdk-nodejs';
 import { Argv } from 'yargs';
 import { RootCommand } from '..';
 
-import { oktaUserAsUser, User, getUser } from './services/user-service';
-import { oktaManageClient, OktaConfiguration } from './services/client-service';
+import { OktaUserService, User, UserService } from './services/user-service';
+import { oktaManageClient } from './services/client-service';
+import * as TE from 'fp-ts/lib/TaskEither';
+import * as E from 'fp-ts/lib/Either';
+import { pipe } from 'fp-ts/lib/function';
+import * as Console from 'fp-ts/lib/Console';
 
-const deactivateUser = async (
-  oktaConfiguration: OktaConfiguration,
+const deactivateUser = (
+  service: UserService,
   userId: string
-): Promise<User> => {
-  const client = oktaManageClient(oktaConfiguration);
-
-  const maybeOktaUser = await getUser(userId, client);
-
-  // eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
-  const deactivate = async (oktaUser: OktaUser) => {
-    // eslint-disable-next-line functional/no-expression-statement
-    await oktaUser.deactivate({
-      sendEmail: false,
-    });
-
-    const deactivatedOktaUser = await client.getUser(userId);
-
-    return oktaUserAsUser(deactivatedOktaUser);
-  };
-
-  // eslint-disable-next-line functional/functional-parameters
-  const throwOnMissing = () => {
-    // eslint-disable-next-line functional/no-throw-statement
-    throw new Error(`User [${userId}] does not exist. Can not de-activate.`);
-  };
-
-  return maybeOktaUser === undefined
-    ? throwOnMissing()
-    : maybeOktaUser.status === UserStatus.DEPROVISIONED
-    ? oktaUserAsUser(maybeOktaUser)
-    : deactivate(maybeOktaUser);
-};
-
+): TE.TaskEither<string, User> =>
+  pipe(
+    userId,
+    service.getUser,
+    TE.chain((maybeUser) =>
+      pipe(
+        service.validateUserExists(maybeUser, userId),
```

Comment:
Again, this is unnecessary complication - better to use a fold.

Reply:
Though by using validateUserExists, it means that all of the commands rely on the same function for checking if the user exists instead of some using folds and others using validation, by having everything in one location, it makes the code easier to manipulate and test does it not?

Reply:
You would still need to test that you are correctly interpreting `Left` and `Right` in the overall flow.

You are also breaking convention by sometimes treating `Right` as `bad`.

Overall, you have more code that - because it breaks convention - is harder to read than the simpler approach of handling the `Option` directly.

Reply:
replaced with fold